### ISSUE TITLE
redir into a local variable doesn't close properly when starting redir from a different function

### ISSUE THREAD

**Issue by Yilin-Yang** [vim/vim#10616]
### Steps to reproduce

Given:
```vim
" repro.vim
function! FirstRedir()
  let l:first_var = ''
  redir => l:first_var
    silent echo 'first'
  redir end
  return l:first_var
endfunction

function! SecondRedir()
  let l:second_var = ''
  redir => l:second_var
    silent echo 'second'
    silent echo FirstRedir()
  redir end
  return l:second_var
endfunction

echomsg SecondRedir()
```

Run:
```bash
vim -i NONE -u repro.vim
```

Results in:
```
yiliny@DESKTOP:~/redir$ ~/temp/vim/usr/local/bin/vim -i NONE -u repro.vim
Error detected while processing /home/yiliny/redir/repro.vim[18]..function SecondRedir[4]..FirstRedir:
line    2:
E121: Undefined variable: l:second_var

first
Press ENTER or type command to continue
```

### Expected behaviour

Given:
```vim
" expected.vim, like before but all variables are global
function! FirstRedir()
  let g:first_var = ''
  redir => g:first_var
    silent echo 'first'
  redir end
  return g:first_var
endfunction

function! SecondRedir()
  let g:second_var = ''
  redir => g:second_var
    silent echo 'second'
    silent echo FirstRedir()
  redir end
  return g:second_var
endfunction

echomsg SecondRedir()
```

```bash
yiliny@DESKTOP:~/redir$ ~/temp/vim/usr/local/bin/vim -i NONE -u expected.vim
^@second
Press ENTER or type command to continue
```

### Version of Vim

8.2.5157

### Environment

Operating system: Ubuntu 22.04, Linux 5.10.60.1-microsoft-standard-WSL2 #1 SMP, Wed Aug 25 23:20:18 UTC 2021, x86_64 GNU/Linux
Terminal: Windows Terminal v1.13.11432.0, tmux 3.2a
Value of $TERM: xterm-256color
Shell: /bin/bash

### Logs and stack traces

This came up while debugging one of my vim plugins: https://github.com/Yilin-Yang/vim-markbar/issues/33

I've done some debugging with GDB in neovim (where this same issue occurs) and in vim (https://github.com/Yilin-Yang/vim-markbar/issues/33#issuecomment-1165134206 and https://github.com/Yilin-Yang/vim-markbar/issues/33#issuecomment-1166169232, respectively), and I think that there's a deeper issue beyond just "the local variables for `SecondRedir()` aren't accessible from inside `FirstRedir()`".

One of my users (https://github.com/Yilin-Yang/vim-markbar/issues/38#issuecomment-1165953681) saw a stacktrace like:

```
Error detected while processing function <SNR>20_ToggleMarkbar[4]..<SNR>20_OpenMarkbar[2]..markbar#MarkbarController#openMarkbar[1]..markbar#MarkbarModel#updateCurrentAndGlobal[26]..markbar#helpers# GetLocalMarks[9]..function <SNR>20_ToggleMarkbar[4]..<SNR>20_OpenMarkbar[2]..markbar#MarkbarController#openMarkbar[1]..markbar#MarkbarModel#updateCurrentAndGlobal[26]..markbar#helpers#GetLocalMarks: E121: Undefined variable: s:ftype_out
```

Where `GetLocalMarks()` looks like:
```
function! markbar#helpers#GetLocalMarks() abort
    let l:redir_output = ''
    try
        redir => l:redir_output
        silent marks abcdefghijklmnopqrstuvwxyz[]<>'`\"^.(){}
        redir end
        " ...
```

I never use a variable named `s:ftype_out` anywhere in vim-markbar. It looks like vim thinks that a `redir =>` statement from preservim/tagbar (https://github.com/preservim/tagbar/blob/0243b19920a683df531f19bb7fb80c0ff83927dd/autoload/tagbar.vim#L38) is still active.

**Comment by bfrg** [comment]
This is what `:help :redir` says:
https://github.com/vim/vim/blob/4e0fc8956649d3208aeaa1642c5efc44e385d77a/runtime/doc/various.txt#L522-L525

**Comment by Yilin-Yang** [comment]
@bfrg You're right, that's the expected behavior: "redir will close any active redirection before starting redirection to the new target".

The initial redirection should be closed (with `"^@second"` being stored into `l:second_var`) as the next redirection is started, which is what happens with the "expected behavior" example that I gave. That doesn't work properly when redirecting into variables local to another function.

**Comment by brammool** [comment]
Yilin Yang wrote:

> ### Steps to reproduce
> 
> Given:
> ```vim
> " repro.vim
> function! FirstRedir()
>   let l:first_var = ''
>   redir => l:first_var
>     silent echo 'first'
>   redir end
>   return l:first_var
> endfunction
> 
> function! SecondRedir()
>   let l:second_var = ''
>   redir => l:second_var
>     silent echo 'second'
>     silent echo FirstRedir()
>   redir end
>   return l:second_var
> endfunction
> 
> echomsg SecondRedir()
> ```
> 
> Run:
> ```bash
> vim -i NONE -u repro.vim
> ```
> 
> Results in:
> ```
> ***@***.***:~/redir$ ~/temp/vim/usr/local/bin/vim -i NONE -u repro.vim
> Error detected while processing /home/yiliny/redir/repro.vim[18]..function SecondRedir[4]..FirstRedir:
> line    2:
> E121: Undefined variable: l:second_var
> 
> first
> Press ENTER or type command to continue
> ```

The ":redir" command does not nest.  What happens here is that when the
second ":redir" is encountered the previous redirection is ended.  But
since the scope is now a different function, "l:second_var" is assigned
to in the wrong function scope.  Since this is done with an append, and
the variable doesn't exist here, you get an error message.

It would be better to give an error that the scope is wrong, but that is
not so easy.  When using a global variable it would still work (but
probably give different results than what you expected).

Making :redir work nested is not that easy.  And I don't think it is
generally useful.

-- 
If someone questions your market projections, simply point out that your
target market is "People who are nuts" and "People who will buy any damn
thing".  Nobody is going to tell you there aren't enough of those people
to go around.
				(Scott Adams - The Dilbert principle)

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
